cranelift/codegen/src/machinst/reg.rs

```diff
@@ -0,0 +1,432 @@
+        if reg.to_reg().to_virtual_reg().is_some() {
+            self.add(Operand::reg_reuse_def(reg.to_reg().to_vreg(), idx));
+        } else {
+            // Sometimes destination registers that reuse a source are
+            // given with RealReg args. In this case, we assume the
+            // creator of the instruction knows what they are doing
+            // and just emit a normal def to the pinned vreg.
+            self.add(Operand::reg_def(reg.to_reg().to_vreg()));
+        }
+    }
+
+    /// Add a register use+def, or "modify", where the reg must stay
+    /// in the same register on the input and output side of the
+    /// instruction.
+    pub fn reg_mod(&mut self, reg: Writable<Reg>) {
+        self.add(Operand::new(
+            reg.to_reg().to_vreg(),
+            regalloc2::OperandConstraint::Reg,
+            regalloc2::OperandKind::Mod,
+            regalloc2::OperandPos::Early,
+        ));
+    }
+}
+
+/// Use an OperandCollector to count the number of operands on an instruction.
+pub fn count_operands<I: MachInst>(inst: &I) -> usize {
+    let mut ops = vec![];
+    let mut coll = OperandCollector::new(&mut ops, |vreg| vreg);
+    inst.get_operands(&mut coll);
+    let (start, end) = coll.finish();
+    debug_assert_eq!(0, start);
+    end as usize
+}
+
+/// Pretty-print part of a disassembly, with knowledge of
+/// operand/instruction size, and optionally with regalloc
+/// results. This can be used, for example, to print either `rax` or
+/// `eax` for the register by those names on x86-64, depending on a
+/// 64- or 32-bit context.
+pub trait PrettyPrint {
+    fn pretty_print(&self, size_bytes: u8, allocs: &mut AllocationConsumer<'_>) -> String;
+
+    fn pretty_print_default(&self) -> String {
+        self.pretty_print(0, &mut AllocationConsumer::new(&[]))
+    }
+}
+
+/// A consumer of an (optional) list of Allocations along with Regs
+/// that provides RealRegs where available.
+///
+/// This is meant to be used during code emission or
+/// pretty-printing. In at least the latter case, regalloc results may
+/// or may not be available, so we may end up printing either vregs or
+/// rregs. Even pre-regalloc, though, some registers may be RealRegs
+/// that were provided when the instruction was created.
+///
+/// This struct should be used in a specific way: when matching on an
+/// instruction, provide it the Regs in the same order as they were
+/// provided to the OperandCollector.
+#[derive(Clone)]
+pub struct AllocationConsumer<'a> {
+    allocs: std::slice::Iter<'a, Allocation>,
+}
+
+impl<'a> AllocationConsumer<'a> {
+    pub fn new(allocs: &'a [Allocation]) -> Self {
+        Self {
+            allocs: allocs.iter(),
+        }
+    }
+
+    pub fn next(&mut self, pre_regalloc_reg: Reg) -> Reg {
+        let alloc = self.allocs.next();
+
+        match (pre_regalloc_reg.to_real_reg(), alloc) {
+            (Some(rreg), _) => rreg.to_reg(),
```

Comment:
Should this debug assert that the allocation matches the real reg?

Reply:
Yes, good point, done!